### PR TITLE
Migrate tests to JUnit5

### PR DESCRIPTION
**Pull request by strangelookingnerd**
There are already some tests that already use JUnit5 as well as some pending PRs (#3219, #3218, ...) to migrate more. This PR will migrate all (*) tests to JUnit5.

Migration includes:

* Using JUnit5 annotations
* Using `Assertions` 
* Remove `public` modifiers from test classes and methods
* Clean up assertions (switching `expected` and `actual`, simplyifing assertions...)
* Trivial code cleanup

I validated my changes using `./mvnw verify` and 

(*) Some tests use `org.junit.Rule` implementations that do not have a JUnit5 replacement and implementing these is non-trivial from what I can tell. Further there are some instances of other testing frameworks that do not yet support JUnit5. I plan on looking into these cases as well but deliberately did not include them in this PR.

## Checklist

* Base your changes on `2.x` branch if you are targeting Log4j 2; use `main` otherwise
* `./mvnw verify` succeeds (if it fails due to code formatting issues reported by Spotless, simply run `./mvnw spotless:apply` and retry)
* Non-trivial changes contain an entry file in the `src/changelog/.2.x.x` directory
* Tests for the changes are provided
* [Commits are signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) (optional, but highly recommended)

### CODE REVIEW

**Comment by ppkarwasz** [comment]
@strangelookingnerd,

It will take me some time to scroll through all the changes in this PR, but I'll try to review this and your other PRs by the end of next week.

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/async/AsyncWaitStrategyFactoryConfigTest.java
Comment (strangelookingnerd) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/async/AsyncWaitStrategyFactoryConfigTest.java:1]:
```
[ERROR] org.apache.logging.log4j.core.async.AsyncWaitStrategyFactoryConfigTest.testIncorrectWaitStrategyFallsBackToDefault(ListAppender, LoggerContext) -- Time elapsed: 5.378 s <<< FAILURE! 
org.opentest4j.AssertionFailedError: expected: <org.apache.logging.log4j.core.async.TimeoutBlockingWaitStrateg>; but was: <com.lmax.disruptor.TimeoutBlockingWaitStrategy>
```

No idea how the changes may be causing this. Could this be some weird class loading issue?

Comment (ppkarwasz) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/async/AsyncWaitStrategyFactoryConfigTest.java:1]: This is probably a problem with dependencies: `com.lmax:disruptor` version 4.x has a `com.lmax.disruptor.TimeoutBlockingWaitStrategy` class, but `com.lmax:disruptor` version 3.x doesn't.

We have two Surefire runs to test asynchronous logger support against both Disruptor 3.x (which supports our target Java 8) and Disruptor 4.x (which requires Java 11).

Comment (ppkarwasz) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/async/AsyncWaitStrategyFactoryConfigTest.java:1]: The actual test results seems OK! Probably the test was broken and you fixed it! :100:

You just need to adapt the expected value of the test to `org.apache.logging.log4j.core.async.TimetoutBlockingWaitStrategy` if `Disruptorutil.DISRUPTOR_MAJOR_VERSION` returns 3 and `com.lmax.disruptor.TimetoutBlockingWaitStrategy` if the constant is 4.

Comment (strangelookingnerd) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/async/AsyncWaitStrategyFactoryConfigTest.java:1]: Added a check in [94884a7](https://github.com/apache/logging-log4j2/pull/3221/commits/94884a70323a634f42ccc92502bfb2396bd21656)

**Comment by ppkarwasz** [comment]
@strangelookingnerd,

Do you have a deadline for the "Migrate to JUnit 5" task? We are swamped with work these days, so the review might take more time than anticipated.

**Comment by strangelookingnerd** [comment]
> @strangelookingnerd,
> 
> Do you have a deadline for the "Migrate to JUnit 5" task? We are swamped with work these days, so the review might take more time than anticipated.

Not at all. I fully understand that the changeset is quite large and will take its time to review. I still hope that it will be merged eventually and I will try my best to keept it up-to-date and free of merge confilict 👍🏼

**Comment by jvz** [comment]
I'll review this, especially since converting tests from v4 to v5 was one of my "hobbies" in this project not that long ago.

**Review by jvz — APPROVED** [review]
Had to review this in IntelliJ as the web UI can barely handle this. Unfortunately, I can't approve it from IntelliJ due to some Apache org settings. So here I am back in the web UI to say LGTM.

**Pull request merged**